### PR TITLE
WIP: Parcel 2 server

### LINKED ISSUE #2144: Parcel 2: Server
The Server from Parcel 1 should be extracted into a separate package in the monorepo for Parcel 2, as `@parcel/server`. This should include both the dev web server as well as the HMR web socket server. All utils related only to servers should be moved into this package. Parcel 1 should also be updated to use this package.

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

I did some fiddling around with the server and hmr socket for v2
The Server was fairly straightforward and should work pretty much entirely now.

HMR on the other hand is probably gonna need a lot of tweaking due to how assets work in Parcel 2?

I could probably use some guidance on how to do this with the asset graph as it probably needs some kind of diff between previous state and new state, which is not available atm afaik?
I currently implemented some kind of state on hmrServer which gets pushed to the client on `buildEnd` which is just a new event I added to AssetGraphBuilder. It doesn't seem to me like the proper way to do this.

I also added a helper function for normalising cli options, I think this will simplify using these options in multiple places a little?

Closes #2144

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [ ] Filled out test instructions (In case there aren't any unit tests)
- [ ] Included links to related issues/PRs

### CODE REVIEW

File: packages/core/core/src/AssetGraphBuilder.js [hunk 1]
```diff
@@ -166,11 +179,27 @@ export default class AssetGraphBuilder extends EventEmitter {
       }
     }
 
+    if (this.hmrServer) {
+      this.hmrServer.updateAsset(cacheEntry);
```
Comment (devongovett) [inline packages/core/core/src/AssetGraphBuilder.js]: A cache entry can contain multiple assets, so probably need to split them out here?

File: packages/core/core/src/AssetGraphBuilder.js [hunk 2]
```diff
@@ -71,6 +73,10 @@ export default class AssetGraphBuilder extends EventEmitter {
         }
       });
     }
+
+    if (opts.cliOpts.hot) {
+      this.hmrServer = new HMRServer(this, opts.cliOpts);
```
Comment (devongovett) [inline packages/core/core/src/AssetGraphBuilder.js]: I am wondering about potentially keeping the HMRServer completely separate from the AssetGraphBuilder, and keep that just for building the graph. Possibly it could live in Parcel instead, which is the glue code between all the pieces?

File: packages/core/core/src/Parcel.js [hunk 3]
```diff
@@ -25,20 +29,39 @@ type ParcelOpts = {
...
     this.options = options;
     this.entries = Array.isArray(entries) ? entries : [entries];
     this.rootDir = getRootDir(this.entries);
+
+    this.normaliseCliOptions();
+  }
+
+  normaliseCliOptions() {
+    this.options.cliOpts.cacheDir =
+      this.options.cliOpts.cacheDir || DEFAULT_CACHE_DIR;
```
Comment (devongovett) [inline packages/core/core/src/Parcel.js]: I kinda liked doing this normalization in the individual places rather than all in one place... What do you think? Why did you move them here?

**Comment by DeMoorJasper** [comment]
Gonna write a new PR for this as this has gotten very out of sync with v2
